pkg/local-gadget-manager/local-gadget-manager.go

```diff
@@ -0,0 +1,409 @@
+	ret := make(map[string]uint64)
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		ret[c.Name] = c.Mntns
+	}
+	return ret
+}
+
+func (l *LocalGadgetManager) LookupPIDByContainer(namespace, pod, container string) uint32 {
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		if container != c.Name {
+			continue
+		}
+		return c.Pid
+	}
+	return 0
+}
+
+func (l *LocalGadgetManager) LookupPIDByPod(namespace, pod string) map[string]uint32 {
+	ret := make(map[string]uint32)
+	for _, c := range l.containers {
+		if namespace != c.Namespace {
+			continue
+		}
+		if pod != c.Podname {
+			continue
+		}
+		ret[c.Name] = c.Pid
+	}
+	return ret
+}
+
+func (l *LocalGadgetManager) GetContainersBySelector(containerSelector *pb.ContainerSelector) []pb.ContainerDefinition {
+	// TODO
+	return []pb.ContainerDefinition{}
+}
+
+// Subscribe returns the list of existing containers and registers a callback
+// for notifications about additions and deletions of containers
+func (l *LocalGadgetManager) Subscribe(key interface{}, selector pb.ContainerSelector, f pubsub.FuncNotify) []pb.ContainerDefinition {
+	l.pubsub.Subscribe(key, func(event pubsub.PubSubEvent) {
+		if match.ContainerSelectorMatches(&selector, &event.Container) {
+			f(event)
+		}
+	})
+	ret := []pb.ContainerDefinition{}
+	for _, c := range l.containers {
+		if match.ContainerSelectorMatches(&selector, &c) {
+			ret = append(ret, c)
+		}
+	}
+	return ret
+}
+
+// Unsubscribe undoes a previous call to Subscribe
+func (l *LocalGadgetManager) Unsubscribe(key interface{}) {
+	l.pubsub.Unsubscribe(key)
+}
+
+func (l *LocalGadgetManager) AddContainer(notif runcfanotify.ContainerEvent) {
+	pid := int(notif.ContainerPID)
+
+	cgroupPathV1, cgroupPathV2, err := containerutils.GetCgroupPaths(pid)
+	if err != nil {
+		return
+	}
+	cgroupPathV2WithMountpoint, _ := containerutils.CgroupPathV2AddMountpoint(cgroupPathV2)
+	cgroupId, _ := containerutils.GetCgroupID(cgroupPathV2WithMountpoint)
+	mntns, err := containerutils.GetMntNs(pid)
+
+	containerDefinition := pb.ContainerDefinition{
+		Id:         notif.ContainerID,
+		Pid:        notif.ContainerPID,
+		Namespace:  "local",
+		Podname:    notif.ContainerID,
+		CgroupPath: cgroupPathV2WithMountpoint,
+		CgroupId:   cgroupId,
+		Mntns:      mntns,
+		CgroupV1:   cgroupPathV1,
+		CgroupV2:   cgroupPathV2,
+	}
+
+	if l.dockercli != nil {
+		filter := filters.NewArgs()
+		filter.Add("id", notif.ContainerID)
+		containers, err := l.dockercli.ContainerList(context.Background(),
+			dockertypes.ContainerListOptions{
+				All:     true,
+				Filters: filter,
+			})
+		if err != nil {
+			fmt.Printf("Couldn't find docker container: %s", err)
+		}
+		if len(containers) == 1 {
+			if len(containers[0].Names) > 0 {
+				containerDefinition.Podname = strings.TrimPrefix(containers[0].Names[0], "/")
```

Comment:
Why you did not used the `containerDefinition.Name` and the `ContainerFilter.ContainerName` for the filtering?

https://github.com/kinvolk/inspektor-gadget/blob/e2260fae3eae0862f6b58f0d9ed55d94a7b2641a/pkg/local-gadget-manager/local-gadget-manager.go#L116-L117

Or what I missed?

Reply:
In Kubernetes, gadgets like "dns" install a tracer on the network namespace of every pod, but not on every container if the pod has several containers.

To make the gadget "dns" work in local-gadget, I needed to put the container name in the Podname field, otherwise the dns gadget would not iterate over each container.

I could change `containerDefinition` to have a new field "NetworkNamespaceOwner" bool that would be set to true for the "pause" container in the Kubernetes world (unless they have `HostNetwork=true` in their PodSpec) and true for all Docker containers with local-kube.... in this way, the dns gadget would check that `NetworkNamespaceOwner` field, and we could stop local-gadget from misusing the `Podname` field.